ivy/functional/backends/paddle/device.py

```diff
@@ -19,62 +21,104 @@
-    )
-    return ret
+    if device is not None:
+        native_dev = as_native_dev(device)
+        if not dev(x, as_native=True)._equals(native_dev):  # type: ignore
+            # TODO: check memory leak because ret is copying the original tensor
+            return paddle.to_tensor(x, place=native_dev, stop_gradient=x.stop_gradient)
+    return x
 
 
-def as_ivy_dev(device: Place, /):
+def as_ivy_dev(device, /):
     if isinstance(device, str):
         return ivy.Device(device)
 
-    if device.is_cpu_place():
-        dev_type = "cpu"
-        dev_idx = 0
-    elif device.is_gpu_place():
-        dev_type = "gpu"
-        dev_idx = device.gpu_device_id()
-
-    if dev_type == "cpu":
-        return ivy.Device(dev_type)
-
-    return ivy.Device(dev_type + (":" + (str(dev_idx) if dev_idx is not None else "0")))
+    if is_native_dev(device):
+        if not isinstance(device, Place):
+            native_dev = Place()
+            native_dev.set_place(device)
+        else:
+            native_dev = device
+        if native_dev.is_cpu_place():
+            return ivy.Device("cpu")
+        elif native_dev.is_gpu_place():
+            dev_type = "gpu"
+            dev_idx = device.gpu_device_id()
+            return ivy.Device(dev_type + ":" + str(dev_idx))
+        elif native_dev.is_gpu_pinned_place():
+            return ivy.Device("gpu:0")  # simiplification
+        else:
+            raise ivy.utils.exceptions.IvyError(
+                f"Cannot convert {device} to an ivyDevice. ",
+                "Currently only cpu, tpu and gpu are supported.",
+            )
+
+    raise ivy.utils.exceptions.IvyError(
+        f"Cannot convert {device} to an ivy device. Expected a "
+        f"paddle.fluid.libpaddle.Place or str, got {type(device)}"
+    )
 
 
 def as_native_dev(
-    device: Optional[Union[ivy.Device, Place]] = None,
+    device: Union[ivy.Device, str, Place],
     /,
 ) -> Optional[Place]:
-    if not isinstance(device, str):
+    if is_native_dev(device):
         return device
-    elif "cpu" in device:
-        return paddle.fluid.libpaddle.CPUPlace()
-    elif "gpu" in device:
-        return paddle.fluid.libpaddle.CUDAPlace()
+    if isinstance(device, str):
+        if "tpu" in device:
+            raise ivy.utils.exceptions.IvyBackendException(
+                "TPU not supported in paddle backend. Consider using ",
+                "TensorFlow or JaX.",
+            )
+        else:
+            # Internal function handles more edge cases and devices
+            _get_paddle_place(device)
+    else:
+        raise ivy.utils.exceptions.IvyError(
+            f"{device} cannot be converted to paddle native device."
+        )
+
+
+def is_native_dev(device, /):
+    return isinstance(
+        device,
+        (
+            core.Place,
+            core.XPUPlace,
+            core.CPUPlace,
+            core.CUDAPinnedPlace,
+            core.CUDAPlace,
+            core.IPUPlace,
```

Comment:
As discussed with @djl11 yesterday, we're currently holding off the support for IPUs and XPUs, so `cpu`, `gpu` and `tpu` would be the only devices supported in `ivy` for now. Same applies to the `torch` backend as well

Reply:
I forgot this pr existed. We are not supporting ipu in Ivy, for supporting ivy we need to support it as an `ivy.Device`. We are just verifying if the backend ipu device is a device and allowing them to run properly in the native backend.

Reply:
Yep sure, but we shouldn't allow it to run properly if device was of the `core.XPUPlace` type for e.g.

Reply:
I mean it's a native device. We are just verifying if it's one, we are not supporting it in ivy or converting it.